### PR TITLE
Saml http metadata resolver refactor

### PR DESCRIPTION
**Pull request by MaciejMierzwa**
### Description
Refactor SamlHTTPMetadataResolver to remove of HTTP commons 4 lib
* Category (Enhancement, Refactoring, Maintenance)
* Behavior should stay the same

### Issues Resolved
- Resolves https://github.com/opensearch-project/security/issues/2932

### Testing
No functionality changes, tests refactored a little.

### Check List
- [ ] New functionality includes testing
- [ ] New functionality has been documented
- [ ] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/com/amazon/dlic/auth/http/saml/HTTPMetadataResolver.java
Comment (willyborankin) [inline src/main/java/com/amazon/dlic/auth/http/saml/HTTPMetadataResolver.java:91]: `response1` is misleading :-) lets call it differently

File: src/main/java/com/amazon/dlic/auth/http/saml/HTTPMetadataResolver.java
Comment (willyborankin) [inline src/main/java/com/amazon/dlic/auth/http/saml/HTTPMetadataResolver.java:122]: I think better to use per-defined constants for headers. e.g. `org.apache.http.HttpHeaders`

File: src/main/java/com/amazon/dlic/auth/http/saml/HTTPMetadataResolver.java
Comment (willyborankin) [inline src/main/java/com/amazon/dlic/auth/http/saml/HTTPMetadataResolver.java:125]: Same better to use constants

File: src/main/java/com/amazon/dlic/auth/http/saml/HTTPMetadataResolver.java
Comment (willyborankin) [inline src/main/java/com/amazon/dlic/auth/http/saml/HTTPMetadataResolver.java:137]: same

**Review by willyborankin — COMMENTED** [review]
Thank you @MaciejMierzwa.  Added some comments.

File: src/main/java/com/amazon/dlic/auth/http/saml/HTTPMetadataResolver.java
Comment (MaciejMierzwa) [reply src/main/java/com/amazon/dlic/auth/http/saml/HTTPMetadataResolver.java:91]: Oh yea, it's leftover from refactoring from the deprecated method. Renamed to just "response"

**Comment by peternied** [comment]
@MaciejMierzwa Can this be backported into 2.x - can you take a look if this is possible?  I'm concerned that the Apache Http Client 5 will create some issues in that codebase

**Comment by cwperks** [comment]
@MaciejMierzwa With this PR can these 3 dependencies (https://github.com/opensearch-project/security/blob/main/build.gradle#L579-L581) be changed to `integrationTestImplementation`?

**Comment by MaciejMierzwa** [comment]
@peternied I think it can be backported into 2.x. Currently 2.x doesn't contain dependencies on any of http5 libraries. If the library is allowed I think it's little bit of additional work.

@cwperks [question](https://github.com/opensearch-project/security/pull/3894#issuecomment-1883926286):
I mostly focused on SamlHTTPMetadataResolver class and the surrounding ones. There is still quite a bit of http4 used in implementation, mostly static headers and HTTP codes are imported from http4 library.

**Comment by peternied** [comment]
> @peternied I think it can be backported into 2.x. Currently 2.x doesn't contain dependencies on any of http5 libraries. If the library is allowed I think it's little bit of additional work.

@MaciejMierzwa We can accept these changes in the main branch, we would be in striking distance to remove all http client 4 references.  There might be other issue aside just importing the library [1], but I'd rather move forward with what you have, just want you to be aware the backport might be a little gross.

- https://github.com/opensearch-project/OpenSearch/issues/3351

File: src/main/java/com/amazon/dlic/util/SettingsBasedSSLConfiguratorV5.java
Comment (peternied) [inline src/main/java/com/amazon/dlic/util/SettingsBasedSSLConfiguratorV5.java:49]: Lets not change the name as I don't think the version number was associated with the Apache Http Client.  What do you think?

File: src/main/java/com/amazon/dlic/auth/http/saml/HTTPMetadataResolver.java
Comment (peternied) [inline src/main/java/com/amazon/dlic/auth/http/saml/HTTPMetadataResolver.java:32]: I see code patterns around the logging that are unique and distinct from the security codebase, was this class's contents copied from another implementation?

**Review by peternied — CHANGES_REQUESTED** [review]

File: src/main/java/com/amazon/dlic/auth/http/saml/HTTPMetadataResolver.java
Comment (MaciejMierzwa) [reply src/main/java/com/amazon/dlic/auth/http/saml/HTTPMetadataResolver.java:32]: You're right, I followed suggestion no. 1 from https://github.com/opensearch-project/security/issues/2932#issue-1788296478
Switched to LogManager.getLogger() for uniformity although i noticed package org.slf4j.LoggerFactory is used in one more place: CleanerFactory.class

**Comment by DarshitChanpura** [comment]
> Hi I noticed that checkstyle fails for http5 library. Looks like this whole PR could be reversed
https://github.com/opensearch-project/security/pull/3741/files

If http 5 is necessary for this PR to resolve, we may need to revert back to http 5 but that would leave this [bug](https://github.com/opensearch-project/security/issues/3581) unresolved. If the only reason we are not moving to http 5 is due to the saml metadata resolver, we can revert the mentioned PR to move back to http 5 since this PR migrates the resolver to http 5 as well.

**Comment by peternied** [comment]
> Hi I noticed that checkstyle fails for http5 library. Looks like this whole PR could be reversed https://github.com/opensearch-project/security/pull/3741/files

The checkstyle issue is associated with the use of client5 specific status codes, you can use the broader ones
```
-import·org.apache.hc.core5.http.HttpStatus;
+import·org.apache.http.HttpStatus;
```

**Comment by DarshitChanpura** [comment]
> The checkstyle issue is associated with the use of client5 specific status codes, you can use the broader ones

Curious, because the changes introduced here refactor the usage from http 4 to http 5 and I believe they are necessary for this PR. Correct me if I'm wrong here @MaciejMierzwa . If they are indeed required to be http5 what should be the path forward?

**Comment by DarshitChanpura** [comment]
@MaciejMierzwa seems like the progress on this PR has stalled. Do you still intend to work on this? Maintainers will be closing this PR out soon if there is no activity.

**Comment by MaciejMierzwa** [comment]
Hi @DarshitChanpura sorry, I see that I got mail notifications, must've missed it. Regarding comments from https://github.com/opensearch-project/security/pull/3894#issuecomment-1894057689 onward - I think the idea behind this ticket was to move towards removal of http4 library from the codebase. @cwperks also highlighted in this comment https://github.com/opensearch-project/security/issues/3581#issuecomment-1772716656 that http4 usage was due to opensaml library, causing conflicts.
For me, it looks like if this PR gets backported to 2.x the bug mentioned by Peter in https://github.com/opensearch-project/security/issues/3581 get's resolved. What do you think?

**Comment by peternied** [comment]
> I see code patterns around the logging that are unique and distinct from the security codebase, was this class's contents copied from another implementation?

@MaciejMierzwa We do not merge unattributed code copied from other code bases, this violates the [DCO](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin) - developer certificate of origin.  I am closing out this PR.

**Pull request closed without merging**
